build/env.py

```diff
@@ -0,0 +1,98 @@
+import os
+import shutil
+import subprocess
+import sys
+import sysconfig
+import tempfile
+import types
+
+
+if False:  # TYPE_CHECKING  # pragma: no cover
+    from typing import Dict, Optional, Iterable, Type
+
+
+class IsolatedEnvironment(object):
+    '''
+    Isolated build environment context manager
+
+    Non-standard paths injected directly to sys.path still be passed to the environment.
+    '''
+
+    def __init__(self):  # type: () -> None
+        self._env = {}  # type: Dict[str, Optional[str]]
+
+    def _replace_env(self, key, new):  # type: (str, Optional[str]) -> None
+        if not new:  # pragma: no cover
+            return
+
+        self._env[key] = os.environ.get(key, None)
+        os.environ[key] = new
+
+    def _restore_env(self):  # type: () -> None
+        for key, val in self._env.items():
+            if val is None:
+                os.environ.pop(key, None)
+            else:
+                os.environ[key] = val
+
+    def _get_env_path(self, path):  # type: (str) -> Optional[str]
+        return sysconfig.get_path(path, vars=self._env_vars)
+
+    def __enter__(self):  # type: () -> IsolatedEnvironment
+        self._path = tempfile.mkdtemp(prefix='build-env-')
+        self._env_vars = {
+            'base': self._path,
+            'platbase': self._path,
+        }
+
+        sys_path = sys.path[1:]
+
+        remove_paths = os.environ.get('PYTHONPATH', '').split(os.pathsep)
+
+        if os.name == 'nt':  # pragma: no cover
+            schemes = ['nt', 'nt_user']
+        else:
+            schemes = ['posix_prefix', 'posix_home', 'posix_user']
+
+        for path in ('purelib', 'platlib'):
+            our_path = sysconfig.get_path(path)
+            if our_path:
+                remove_paths.append(our_path)
+
+            for scheme in schemes:
+                our_path = sysconfig.get_path(path, scheme)
+                if our_path:
+                    remove_paths.append(our_path)
+
+            env_path = self._get_env_path(path)
+            if env_path:
+                sys_path.append(env_path)
+
+        for path in remove_paths:
+            if path in sys_path:
+                sys_path.remove(path)
+
+        self._replace_env('PATH', self._get_env_path('scripts'))
```

Comment:
Any reason why you try to alter the current environment, instead of just cloning the current environ, pass that in to subprocess calls?

Reply:
Because we want everything inside the context manager to get this environment, not just the pip call. If we don't alter the environment the `pep517` hook calls won't be isolated for eg.

Reply:
Maybe put these three in a dictionary and write a for loop then to make it easy to read and extend.

Reply:
IMO that's just more code I'll have to read when looking at the code. I will have to see that we are creating a dictionary, remember what is there and then see what we do to the dictionary.

Right now we can see directly what we are doing.

If you have strong feelings about this I can change though, I don't.